### DISCUSSION TITLE
Refresh / update model from partial data.

### DISCUSSION THREAD

**Discussion by jheddings** [pydantic/pydantic#3139] [Question] [answered]
I'm looking for a convenient way to refresh a model from data w/out having to initialize a new object each time.

So, rather than having to (for example):

```python
orig_data = get_current_data(...)

while True:
    new_data = get_data_updates(...)
    data = orig_data | new_data
    item = ItemModel(**data)
    ...
```

This would be preferred:
```python
data = get_current_data(...)
item = ItemModel(**data)

while True:
    data = get_data_updates(...)
    item.update(**data)  # <- some magic method here?
    ...
```

Any pointers would be appreciated!

**Comment by jheddings** [comment]
After poking around a bit, I lifted the following code from the constructor that seems to be working:

```python
    def refresh(__pydantic_self__, **data):
        values, fields, error = validate_model(__pydantic_self__.__class__, data)

        if error:
            raise error

        object.__setattr__(__pydantic_self__, '__dict__', values)
```

Adding this to my model seems to do the trick.  Is this a reasonable approach?

**Comment by jheddings** [comment]
An updated version that seems to work better...

```python
    def refresh(__pydantic_self__, **data):
        """Refresh the internal attributes with new data."""

        values, fields, error = validate_model(__pydantic_self__.__class__, data)

        if error:
            raise error

        for name in fields:
            value = values[name]
            log.debug("set object data -- %s => %s", name, value)
            setattr(__pydantic_self__, name, value)

        return __pydantic_self__
```

I'm still open to any ideas for improving this method!

**Comment by jpribyl** [comment]
I think this could be implemented even more simply as a class method:
```python
class MyClass(BaseModel):
  my_field: str = Field(None)
  
  def update(self, data: Dict) -> MyClass:
    for k,v in self.validate(data).dict().items():
      log.debug(f"updating value of '{k}' from '{getattr(self, k, None)}' to '{v}'")      
      setattr(self, k, v)
    return self
```

using `self.validate` will raise by default :smile:

**Reply by Hrabal** [reply]
My two cents here: keep in mind that this solution will set fields that are not in data to the defaults, and raise for missing required fields, so basically does not "update from partial data" as OP's request, but "resets to partial data", se example code:

``` python
>>> from pydantic import BaseModel
>>>
>>> class Test(BaseModel):
...     required: str
...     optional: str = None
...     def update(self, data):
...         for k,v in self.validate(data).dict().items():
...           setattr(self, k, v)
...         return self
...
>>> test = Test(required="foo", optional="bar")
>>> test
Test(required='foo', optional='bar')
>>> test.update({"required": "baz"})  # Updating only required
Test(required='baz', optional=None)  # Optional goes to default
>>> test.update({"required": None})  # Updating only optional
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 5, in update
  File "pydantic/main.py", line 711, in pydantic.main.BaseModel.validate
  File "pydantic/main.py", line 342, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for Test
required
  none is not an allowed value (type=type_error.none.not_allowed)
```

OP's solution in comment instead updates given fields with no interaction with the missing one.

**Reply by anilsg** [reply]
FYI, there is some discussion on support for partial updates (for PATCH operations) here: https://github.com/pydantic/pydantic/discussions/3089

I also include an implementation of a function that can be used in the path operation function to transform the usual `BaseModel` in use to all-fields-optional, as I think is mentioned in this thread somewhere.

I also note that `BaseModel` already implements `copy` which takes an `update` parameter to apply partial updates to a copy of the model. Tragically it doesn't deal with sub-objects properly but it seems this may be in progress...

**Reply by noranraskin** [reply, marked as answer]
To get around the problems mentioned by @Hrabal (raise on missing values or resetting to default) you can first update a copy and validate that. Alternatively if you do want to raise on missing required values and only want to fix the reset to default, you can add `exclude_defaults` when calling `dict()`. I added both changes down below:
```python
class MyClass(BaseModel):
  my_field: str = Field(None)

  def update(self, data: dict) -> MyClass:
    update = self.dict()
    update.update(data)
    for k,v in self.validate(update).dict(exclude_defaults=True).items():
      log.debug(f"updating value of '{k}' from '{getattr(self, k, None)}' to '{v}'")      
      setattr(self, k, v)
    return self
```

**Reply by ale-hm** [reply]
This doesn't seem to take into account field alias.

**Reply by pycaw** [reply]
What we need are proper official `assign` and `update` methods. Former assigning the contents of another object (`dict`, `BaseModel`, etc) resulting in a validated model instance afterwards, and the latter doing a partial update that is yet to be validated. Or at least that is one way of going about it. A simple `validate`/`validate_self` method is also lacking, I have come across that with other problems as well.

**Reply by dd-ssc** [reply]
@noranraskin: Your code might possibly work perfectly, but your choice of variable names is really unfortunate. Would you mind changing one of the various `update`s to something less confusing ?

**Reply by noranraskin** [reply]
@dd-ssc well you don't have to copy paste my code as is. Feel free to rename anything however you want

**Comment by samuelcolvin** [comment]
I think it short answer here is that this is conceptually wrong.

You should create a new model (or typeddict in v2) for each new request.

Of course you can use `setattr` and it'll work, but my instinct is that it's the wrong route.

**Reply by arogozhnikov** [reply]
I believe that's not how sqlmodel (pydantic + sqlalchemy)  can work. To update record in the db, user should retrieve an object and modify it. In case a new model is passed, it will be considered as attempt to register a new record, not modify existing one

**Reply by samuelcolvin** [reply]
I disagree, if you want to update a record, use a different model with all fields optional, and validate against that. You can add a root validator to confirm that at least one field has been set.

**Reply by arogozhnikov** [reply]
you completely ignored essence of my comment. sqlmodel uses object identity to map to object in the database. Period. You can't use new object for data modification, you should modify the one retrieved by sqlmodel.

See docs: https://sqlmodel.tiangolo.com/tutorial/update/#review-the-code
